Create a GroupingRequest with grouping properties

yjtakamabe opened the pull request:
**Is your feature request related to a problem? Please describe.**
The values from the query profile are not reflected when creating a GroupingRequest.

**Describe the solution you'd like**
We want to create GroupingRequest whose default values are determined by grouping.defaultMaxGroups/ grouping.defaultMaxHits/ grouping.globalMaxGroups/ grouping.defaultPrecisionFactor in the query profile.

**Describe alternatives you've considered**
We have implemented initProperty and longProperty methods in our container to set the default values in the query profile.

```
public Result search(Query query, Execution execution) {
    GroupingRequest request = GroupingRequest.newInstance(query);
    intProperty(query, GroupingQueryParser.PARAM_DEFAULT_MAX_GROUPS).ifPresent(request::setDefaultMaxGroups);
    intProperty(query, GroupingQueryParser.PARAM_DEFAULT_MAX_HITS).ifPresent(request::setDefaultMaxHits);
    longProperty(query, GroupingQueryParser.GROUPING_GLOBAL_MAX_GROUPS).ifPresent(request::setGlobalMaxGroups);
    // ...
    // some logic here
    // ...
}

private static OptionalInt intProperty(com.yahoo.search.Query q, CompoundName name) {
    Integer val = q.properties().getInteger(name);
    return val != null ? OptionalInt.of(val) : OptionalInt.empty();
}

private static OptionalLong longProperty(com.yahoo.search.Query q, CompoundName name) {
    Long val = q.properties().getLong(name);
    return val != null ? OptionalLong.of(val) : OptionalLong.empty();
}
```

**Additional context**
We cannot use the setDefaultPrecisionFactor method in GroupingRequest because it is package-private.

yjtakamabe committed:
Create a GroupingRequest with grouping properties

yjtakamabe commented:
Please let me know if there's anything I can do to help move it forward or if you have any feedback.
Thank you for your time.

bjorncs commented:
Hi @yjtakamabe. Thanks for your contribution. I will publish my review here shortly.

bjorncs commented on container-search/src/main/java/com/yahoo/search/grouping/GroupingRequest.java:
```diff
@@ -188,6 +188,22 @@ public static GroupingRequest newInstance(Query query) {
+     */
+    public static GroupingRequest newInstanceWithGroupingProperties(Query query) {
+        GroupingRequest newRequest = new GroupingRequest(query.getSelect());
+        GroupingQueryParser.intProperty(query, GroupingQueryParser.PARAM_DEFAULT_MAX_GROUPS).ifPresent(newRequest::setDefaultMaxGroups);
+        GroupingQueryParser.intProperty(query, GroupingQueryParser.PARAM_DEFAULT_MAX_HITS).ifPresent(newRequest::setDefaultMaxHits);
+        GroupingQueryParser.longProperty(query, GroupingQueryParser.GROUPING_GLOBAL_MAX_GROUPS).ifPresent(newRequest::setGlobalMaxGroups);
+        GroupingQueryParser.doubleProperty(query, GroupingQueryParser.PARAM_DEFAULT_PRECISION_FACTOR).ifPresent(newRequest::setDefaultPrecisionFactor);
```
I think we can just introduce this property propagation to `GroupingRequest.newInstance()` and not introduce another static factory method. That will also remove the necessity of propagating these parameters in `GroupingQueryParser` (line 77-80). The `{int|long|double}Property` helpers can be moved to `GroupingRequest`.

yjtakamabe replied:
I will proceed in line with your suggestion.
Thank you for the feedback.

bjorncs reviewed (commented):
See comments on `newInstanceWithGroupingProperties`. 

Feel free to change the access modifier of `setDefaultPrecisionFactor` to `public`.

yjtakamabe committed:
Update property propagation in GroupingRequest.newInstance().
Removing extra parameters in GroupingQueryParser.
Move The {int|long|double}Property helpers to GroupingRequest.

bjorncs reviewed (approved):
LGTM 👍

yjtakamabe commented:
Thank you for approving my contribution.
